src/wallet/wallet2.h

```diff
@@ -1927,6 +1942,34 @@ namespace boost
         return;
       a & x.multisig_sigs;
     }
+
+    template <class Archive>
+    inline typename std::enable_if<!Archive::is_loading::value, void>::type serialize(Archive &a, boost::container::flat_map<crypto::public_key, cryptonote::subaddress_index> &x, const boost::serialization::version_type ver)
+    {
+      uint32_t sz = x.size();
+      a & sz;
+      for (const auto &e: x)
+      {
+        a & e.first;
+        a & e.second;
+      }
+    }
+    template <class Archive>
+    inline typename std::enable_if<Archive::is_loading::value, void>::type serialize(Archive &a, boost::container::flat_map<crypto::public_key, cryptonote::subaddress_index> &x, const boost::serialization::version_type ver)
+    {
+      using pair_type = std::pair<crypto::public_key, cryptonote::subaddress_index>;
+      uint32_t sz;
+      a & sz;
+      std::vector<pair_type> vector;
+      vector.resize(sz);
+      for (uint64_t i = 0; i < sz; ++i)
+      {
+        a & vector[i].first;
+        a & vector[i].second;
+      }
+      std::sort(vector.begin(), vector.end(), [](const pair_type &e0, const pair_type &e1){ return e0.first < e1.first; });
```

Comment:
One thing I forgot to mention - serialization load functions typically overwrite the existing container value. This is merging in the values into the existing container. Usage of `adopt_sequence` without the `sort`, would do a replace and uniqueness check in one (I looked at the implementation this time, instead of that insert goof).

Reply:
I mentioned this somewhere else (incorrectly for that chunk of code) - the current implementation for serialization loading is relying on user input to guarantee uniqueness. Using this:
```c++
x.clear();
x.reserve(sz);
pair_type e{};
while (sz--)
{
    a & e.first;
    a & e.second;
    x.insert(x.end(), e);
}
```
will filter out duplicates from user input and is ~27.5% faster (100,000 random public keys in a `std::stringstream` using release/-Ofast with gcc 7.3 and me hopefully negating CPU burstiness weirdness) than the current implementation. The only time the above version would be slower is if an alternative output serialization function put the elements _not_ in-order.

The reason for the performance increase is because if the archive elements are in-order, the loop becomes `vector.insert(vector.end(), e);` internally within Boost (with some checks/branches in case the assumption is invalid). This removes the temporary buffer, sorting/searching, and extra element `memcpy`. `boost::serialization` attempts a similar optimization with `std::map`.

Reply:
OK, that looks nice, I'll push that, thanks.